server/apps/immich/src/api-v1/user/user.domain.ts

```diff
@@ -0,0 +1,80 @@
+import { UserEntity } from '@app/database/entities/user.entity';
+import { BadRequestException, ForbiddenException, InternalServerErrorException, Logger } from '@nestjs/common';
+import { genSalt, hash } from 'bcrypt';
+import { AuthUserDto } from '../../decorators/auth-user.decorator';
+import { CreateAdminDto, CreateUserDto, CreateUserOauthDto } from './dto/create-user.dto';
+import { IUserRepository } from './user-repository';
+
+export class UserDomain {
+  constructor(private userRepository: IUserRepository) {}
+
+  private async generateSalt(): Promise<string> {
+    return genSalt();
+  }
+
+  private async hashPassword(password: string, salt: string): Promise<string> {
+    return hash(password, salt);
+  }
+
+  async updateUser(authUser: AuthUserDto, user: UserEntity, update: Partial<UserEntity>): Promise<UserEntity> {
+    if (!authUser.isAdmin && (authUser.id !== user.id || user.id != update.id)) {
+      throw new ForbiddenException('You are not allowed to update this user');
+    }
+
+    try {
+      const payload: Partial<UserEntity> = { ...update };
+      if (payload.password) {
+        const salt = await this.generateSalt();
+        payload.salt = salt;
+        payload.password = await this.hashPassword(payload.password, salt);
+      }
+      return this.userRepository.update(user.id, payload);
+    } catch (e) {
+      Logger.error(e, 'Failed to update user info');
+      throw new InternalServerErrorException('Failed to update user info');
+    }
+  }
+
+  async createUser(createUserDto: CreateUserDto | CreateAdminDto | CreateUserOauthDto): Promise<UserEntity> {
+    const user = await this.userRepository.getByEmail(createUserDto.email);
+    if (user) {
+      throw new BadRequestException('User exists');
+    }
+
+    if (!(createUserDto as CreateAdminDto).isAdmin) {
```

Comment:
This is part of the business logic. Checks if we are creating a admin user (i.e. we are passing a `CreateAdminDto`):
I could use a TypeScript narrowing function if is easier to understand. Something like this:
```typescript
function isCreateUserDto(createUserDto: CreateUserDto | CreateAdminDto | CreateUserOauthDto): createUserDto is CreateUserDto {
    return (createUserDto as CreateAdminDto).isAdmin === true
  }
```
I think is better if we have all the business logic related to how we create a user in the same place, instead of scattered in services that use this core domain file.

Reply:
The only business logic here is that "the first account created needs to be an admin", it's not enforcing what specific dto was sent. In the future if the OAuthDto sent `isAdmin: true` this code would execute the same way, even though the way we use types here is wrong/misleading. So, I don't think the types used here or the type casting make sense, since they don't actually align with the business logic that we're validating.